File: data_structures/queues/queue_by_list.js
/**
 * Queue implementation in JavaScript
 */
class Queue {
    constructor() {
        this.entries = [];
    }

    put(item) {
        this.entries.push(item);
    }

    get() {
        if (this.entries.length === 0) {
            throw new Error("Queue is empty");
        }
        return this.entries.shift();
    }

    getFront() {
        if (this.entries.length === 0) {
            throw new Error("Queue is empty");
        }
        return this.entries[0];
    }

    rotate(rotation) {
        for (let i = 0; i < rotation; i++) {
            this.put(this.get());
        }
    }

    size() {
        return this.entries.length;
    }

    isEmpty() {
        return this.entries.length === 0;
    }
}

const q = new Queue();
q.put(10);
q.put(20);
q.put(30);
console.log("Front:", q.getFront());
q.rotate(1);
console.log("Front after rotation:", q.getFront());
console.log("Get:", q.get());

File: data_structures/queues/queue_by_list.go
package main

import (
	"errors"
	"fmt"
)

type Queue struct {
	entries []interface{}
}

func (q *Queue) Put(item interface{}) {
	q.entries = append(q.entries, item)
}

func (q *Queue) Get() (interface{}, error) {
	if len(q.entries) == 0 {
		return nil, errors.New("Queue is empty")
	}
	item := q.entries[0]
	q.entries = q.entries[1:]
	return item, nil
}

func (q *Queue) GetFront() (interface{}, error) {
	if len(q.entries) == 0 {
		return nil, errors.New("Queue is empty")
	}
	return q.entries[0], nil
}

func (q *Queue) Rotate(rotation int) {
	if len(q.entries) == 0 {
		return
	}
	for i := 0; i < rotation; i++ {
		item, _ := q.Get()
		q.Put(item)
	}
}

func (q *Queue) Size() int {
	return len(q.entries)
}

func main() {
	q := &Queue{}
	q.Put(10)
	q.Put(20)
	q.Put(30)
	
	val, _ := q.GetFront()
	fmt.Println("Front:", val)
	
	q.Rotate(1)
	val, _ = q.GetFront()
	fmt.Println("Front after rotation:", val)
	
	val, _ = q.Get()
	fmt.Println("Get:", val)
}
